feat(DPGO-19, 21): email sender, password restore story

glbter opened the pull request:
- gmail oauth email sender with account dp210go
- password restore feature
- change password feature

glbter committed:
feat(DPGO-21): basic oauth gmail api email sender

glbter committed:
chore(DPGO-21): password code sender

glbter committed:
feat(DPGO-19): add send password code endpoint

glbter committed:
chore(DPGO-19): passw code repo

glbter committed:
chore(DPGO-19): passw cases with cache, auth func

also, add userRepo if exists by email, auth func authenticates with restore code

cache is used, as we planned to store codes in cache

glbter committed:
feat(DPGO-19): authenticate via restore code api

glbter committed:
Merge branch 'dev' into password-restore-dev-merge

# Conflicts:
#	README.md
#	go.mod
#	go.sum

glbter committed:
chore(DPGO-19): remove pasw code repo

cache will be used

glbter committed:
chore(DPGO-21): merge config, gmail instuction

igkostyuk commented on internal/server/http/user/handlers.go:
```diff
@@ -26,18 +26,25 @@ type UsersUsecases interface {
-	usecases UsersUsecases
-	logger   *zap.Logger
+	userCases UsersUsecases
+	paswCases PasswordUsecases
```
why do we need  it here ? can we have separate handlers ?

glbter replied:
@igkostyuk I will think about it. If we are going to use it as other service, I will move it out soon.

igkostyuk replied:
for this change user handlers can't exist without paswCases

igkostyuk commented on internal/repository/postgres/user/repo.go:
```diff
@@ -114,6 +114,16 @@ func (r *Repository) GetByEmail(ctx context.Context, email string) (entity.User,
 	return u, nil
 }
 
+func (r *Repository) UserExists(ctx context.Context, email string) bool {
```
you can take user by email, and check error

glbter replied:
@igkostyuk do we need to take a user from a db, or our user would be lightwight as it is now?
As this aproach uses the advantages of using SQL queries over ORMs

igkostyuk replied:
lightweight user ? do you worry about performance issues? do we need user exist by id ?

igkostyuk replied:
do we need userExist by id ? do we feel any performance issues ? lightwight user ?

glbter committed:
feat(DPGO-19): change password endpoint

glbter commented:
@glbter need to fix change password: forgot to save new hash

igkostyuk commented on internal/entity/user.go:
```diff
@@ -13,9 +13,19 @@ type User struct {
+}
+
+type PasswordsRequest struct {
+	UserID          string `json:"id" validate:"omitempty"`
```
in user with ID  userID :smile: 
I need to do something with user entity, right now it's too much inside it, maybe i will move it to handlers.

glbter replied:
Heh, actually, I need only id from user there. I did not embed it.
I thought about Roman's idea. Possible solution is to store passwords separately

igkostyuk replied:
not sure how to do it right, need some research and rewrite it couple of times and test

igkostyuk replied:
I will do something like this https://threedots.tech/post/common-anti-patterns-in-go-web-applications/

glbter replied:
@igkostyuk As I understood, your idea is to use less embedding but to make several entities. Or you talk about dto and entities?

glbter committed:
fix(DPGO-19): save new password

glbter committed:
fix: don't return or update password with user data

this data should be secured, it shouldn't be returned or updated with other data

glbter committed:
refactor(DPGO-19): divide handlers into topics

user and password topic, generic util funcs and validator

glbter committed:
refactor: rename to descriptible names

igkostyuk commented on internal/server/http/handlers/user/handlers.go:
```diff
@@ -0,0 +1,137 @@
+package user
```
can you explain why do delet my file and replace it with yours ?

glbter replied:
@igkostyuk I did not. that is the thing. Just moved it to a descriptive package - handlers

igkostyuk replied:
yes you are. I lose all history.

glbter replied:
ok, if it is a trouble for u, we will leave this file out

igkostyuk replied:
it's not only for me

glbter replied:
I don't know what are you going to achive... It is ok for you to leave it where you created it?

igkostyuk replied:
you just delete 192 lines without a reason, i will show you in monday. you can check diff

glbter replied:
@igkostyuk Я правда не понимаю к чему спор.  Причину я обьяснил. Я не против оставить как было и сделать по твоему.

igkostyuk replied:
не удаляй файлы двигай их. в истории непонятно что ты там менял как ты менял.. ты все мои хендлеры удалил и заменил их на свои.
